test/fs.js

```diff
@@ -0,0 +1,186 @@
+'use strict';
+
+const metatests = require('metatests');
+const metasync = require('metasync');
+const fs = require('fs');
+const path = require('path');
+const stream = require('stream');
+const { copyDir } = require('..');
+
+const testDataFile =
+  'JySAF1figbONWxAeeoGhDXeI7df25GKXKPcyCl63ZjIv5HZGQH8yfFeHGGaSJ9zDMP0CwtgX2madIQkFmuq38SNhcKcGQ';
+const srcDir = 'sourceDirectory';
+const targetDir = 'targetDirectory';
+const absoluteLink = __dirname.replace('test', 'lib');
+const files = [
+  '1.js',
+  '2.txt',
+  '3.html',
+  '4.css',
+  '5.doc',
+  '6.tmp',
+  '7.asm',
+  '8.zip',
+  'link',
+  'dir1',
+  'dir2',
+  'dir3',
+];
+
+const createTestDir = (current, cb, curDepth = 1) => {
+  if (curDepth > 4) {
+    cb();
+    return;
+  }
+  fs.mkdir(current, err => {
+    if (err) {
+      cb(err);
+      return;
+    }
+    metasync.each(
+      files,
+      (file, cb) => {
+        const srcFile = path.join(current, file);
+        if (file.includes('.')) {
+          fs.appendFile(srcFile, testDataFile, cb);
+        } else if (file === 'link') {
+          fs.symlink(absoluteLink, path.join(current, file), cb);
+        } else {
+          createTestDir(srcFile, cb, curDepth + 1);
+        }
+      },
+      cb
+    );
+  });
+};
+
+const rmdirp = (dir, cb) => {
+  fs.lstat(dir, (err, stats) => {
+    if (err) {
+      if (err.code === 'ENOENT') cb(null);
+      else cb(err);
+      return;
+    }
+    if (stats.isDirectory()) {
+      fs.readdir(dir, (err, files) => {
+        if (err) {
+          cb(err);
+          return;
+        }
+        files = files.map(file => path.join(dir, file));
+        metasync.each(files, rmdirp, err => {
+          if (err) cb(err);
+          else fs.rmdir(dir, cb);
+        });
+      });
+    } else {
+      fs.unlink(dir, cb);
+    }
+  });
+};
+
+class ToUpperCase extends stream.Transform {
+  constructor(options = {}) {
+    options = Object.assign(options, {
+      decodeStrings: false,
+    });
+    super(options);
+  }
+  _transform(chunk, encoding, callback) {
+    if (encoding !== 'utf8') {
+      this.emit('error', new Error('only utf8'));
+      return callback();
+    }
+    this.push(chunk.toUpperCase());
+    return callback();
+  }
+}
+
+metatests.test('fs / copyDir', test => {
+  createTestDir(srcDir + '1', err => {
```

Comment:
It will be better to use `beforeEach` and `afterEach` to avoid boilerplate code.

Reply:
Again question about ```beforeEach``` and ```afterEach```, are these something from metatests?

Reply:
Yes, you can learn about it more [here](https://github.com/metarhia/metatests/blob/d5fbe7827bd388d9d78c2bb7f98cd02b83331562/lib/imperative-test.js) and [here](https://github.com/metarhia/filestorage/blob/master/test/filestorage.js). It will allow executing the function before and after the specified set of tests e.i. create test files and clean up them.

Reply:
I have read about this and tested, but I have 6 different dirs names and I can't just use one beforeEach+afterEach for all the, because I should change dir name srcDir + 'x'
or I can do my tests consistently

``` javascript
fsTest.beforeEach((test, cb) => {
  createTestDir(srcDir + '1', err => {
    if (err) {
      test.error(err);
      return;
    }
    cb();
  });
});
fsTest.afterEach((test, cb) => {
  rmdirp(srcDir + '1', err => {
    if (err) {
      test.error(err);
      return;
    }
    rmdirp(targetDir + '1', err => {
      if (err) {
        test.error(err);
        return;
      }
      cb();
    });
  });
});
```

Reply:
You need only one test dir for each test. `beforeEach` and `afterEach` are created specifically for this. Also, about `createTestDir`. Why are you trying to use recursion so much when it can be easily avoided? Tests don't need to have high performance, they should be simple enough so they can be inspected in the future and they need to be an example of usage. Something like this is closer to what we need.
```js
const srcFiles = [
  'test/srcDir',
  'test/srcDir/file1',
  'test/srcDir/dir1',
  'test/srcDir/dir2',
  'test/srcDir/dir1/subdir1',
];

const dstFiles = [
  'test/dstDir',
  'test/dstDir/file1',
  'test/dstDir/dir1',
  'test/dstDir/dir2',
  'test/dstDir/dir1/subdir1',
];

const copydirTest = metatests.test('Recursive dir copy');
copydirTest.endAfterSubtests();

copydirTest.beforeEach((test, cb) => {
  metasync.series(srcFiles, (file, cb) => {
      if (path.basename(file).startsWith('file')) {
        fs.appendFile(file, testData, cb);
      } else {
        fs.mkdir(file, cb);
      }
    }, test.cb(cb)
  );
});

copydirTest.afterEach((test, cb) => {
  metasync.series(common.merge(srcFiles, dstFiles).reverse(), (file, cb) => {
      if (path.basename(file).startsWith('file')) {
        fs.unlink(file, cb);
      } else {
        fs.rmdir(file, cb);
      }
    }, test.cb(cb)
  );
});

copydirTest.test('subtest 1', test => {
  ...
}

copydirTest.test('subtest 2', test => {
  ...
}
```
To be clear: this is not the only way to do this, that is just an example, but it is more readable and more efficient than the existing implementation. If you have questions about how [@metarhia/metatests](https://github.com/metarhia/metatests) are used you can check out existing tests in this or other repositories.

Reply:
Yes, I just don't know the whole potential/usage of metatests.